cmd/maya-apiserver/spc-actions/select_disk.go

```diff
@@ -0,0 +1,207 @@
+// with the pool constraints.
+
+// NOTE: Not all the selected nodes may qualify.
+
+// Finally diskSelector function will vote for qualified nodes.
+
+func nodeSelector(listDisk *v1alpha1.DiskList, poolType string) map[string]*nodeDisk {
+
+	// nodeDiskMap is the data structure holding host name as key
+	// and nodeDisk struct as value
+	nodeDiskMap := make(map[string]*nodeDisk)
+	for _, value := range listDisk.Items {
+		// if no more allotment is required, stop processing
+		if spareAllotment == 0 {
+			glog.Info("required pool allotment done")
+			break
+		}
+		if nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]] == nil {
+			// Entry to this block means first time the hostname will be mapped for the first time.
+			// Obviously, this entry of hostname(node) is for a usable disk and initialize diskCount to 1.
+			nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]] = &nodeDisk{diskList: []string{value.Name}, diskCount: 1}
+			// If pool type is striped the node qualifies for pool creation hence spareAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+				spareAllotment--
+			}
+		} else {
+			// Entry to this block means the hostname was already mapped and it has more than one disk and at least two disks.
+			nodeDisk := nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]]
+			// Increment the disk count
+			nodeDisk.diskCount++
+			// Add the current disk to the diskList for this node.
+			nodeDisk.diskList = append(nodeDisk.diskList, value.Name)
+			// If pool type is mirrored the node qualifies for pool creation hence spareAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeMirroredCPK) {
+				if nodeDisk.diskCount == int(v1alpha1.MirroredDiskCountCPK) {
+					spareAllotment--
+				}
+			}
+		}
+
+	}
+	return nodeDiskMap
+}
+
+// diskSelector is the function that will select the required number of disks from qualified nodes
+// so as to provision storagepool
+
+func diskSelector(nodeDiskMap map[string]*nodeDisk, poolType string) []string {
+
+	// selectedDisk will hold a list of disk that will be used to provision storage pool, after a
+	// minimum number of node qualifies
+	var selectedDisk []string
+
+	// requiredDiskCount will hold the required number of disk that should be selcted from a qualified
+	// node for specific pool type
+	var requiredDiskCount int
+
+	// Range over the nodeDiskMap map to get the list of disks
+	for _, val := range nodeDiskMap {
+		// If pool type is striped, 1 disk should be selected
+		if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+			requiredDiskCount = int(v1alpha1.StripedDiskCountCPK)
+		}
+		// If pool type is striped, 2 disks should be selected
+		if poolType == string(v1alpha1.PoolTypeMirroredCPK) {
+			requiredDiskCount = int(v1alpha1.MirroredDiskCountCPK)
+			// If the current disk count on the node is less than the required disks
+			// then this is a dirty node and it will not qualify.
+			if len(val.diskList) < requiredDiskCount {
```

Comment:
Should this condition be outside of this if block? Or the assumption is that only nodes with more than 1 disk are present.

Reply:
The block will be executed only if pool type is mirrored and there the assumption is that node with more than 1 usable disk get selected